### ISSUE TITLE
ScalaTest: allow invoking custom logic prior to writing pact files

### ISSUE THREAD

**Issue by solarmosaic-kflorence** [jbwheatley/pact4s#8]
Instead of writing out pacts at the end of `run(testName: Option[String], args: Args)`, it would be more flexible to do it in `afterAll(): Unit` of the `BeforeAndAfterAll` trait -- that way users can provide custom logic to fail the tests prior to the pact files being written.

In our case, I want to ensure that a test exists for every message defined. If not, I want the suite to fail and pacts to not be written.

**Comment by jbwheatley** [comment]
I'm in two minds about this, as it does make the `PactForger` mixin heavier, and mean that using `BeforeAndAfterAll` for anything else on top of this becomes a bit clunkier. Maybe the `PactForger` trait itself could provide a `beforeWritePacts` method or something like that?

**Comment by solarmosaic-kflorence** [comment]
I think it's pretty common in Scalatest to do:

```
override def afterAll(): Unit = {
  // .. do whatever
  super.afterAll()
}
```

But yes, having a method like `beforeWritePacts` would be more explicit. I guess I should have made this a problem statement instead of a solution -- I am really open to any solution for this.

**Comment by jbwheatley** [comment]
True about it being common in scalatest, I just don't like it 😛

Actually it would be good to have it as a pact4s native method, because them we can include it for the other testing Frameworks. I would like to keep feature parity between them all if possible

**Comment by jbwheatley** [comment]
@solarmosaic-kflorence I added a `beforeWritePacts` method to the pact forger traits. For scalatest its got signature `() => Either[Throwable, Unit]`.  If it's a `Left` then the error gets thrown and the pacts don't get written. Available in v0.0.7.

**Comment by solarmosaic-kflorence** [comment]
Looks good, thanks!
